torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -722,133 +715,172 @@ def _step_microbatches(
         total_steps = warmup_steps + fwd_bwd_steps + cooldown_steps
 
         logger.debug(
-            f"rank {self.rank}, warmup_steps {warmup_steps}, "  # noqa: G004
-            f"1f1b {fwd_bwd_steps}, cooldown_steps {cooldown_steps}"
+            "rank %s, warmup_steps %s, 1f1b %s, cooldown_steps %s",
+            rank,
+            warmup_steps,
+            fwd_bwd_steps,
+            cooldown_steps,
         )
 
-        def forward_stage_local_index(step):
-            return (step // self.pp_group_size) % self.n_local_stages
+        # Calculates the stage index based on step and pp_group_size
+        def forward_stage_index(step):
+            # Get the local index from 0 to n_local_stages-1
+            local_index = (step // self.pp_group_size) % self.n_local_stages
+            return (local_index * self.pp_group_size) + rank
 
-        def backward_stage_local_index(step):
-            return (
+        def backward_stage_index(step):
+            local_index = (
                 self.n_local_stages
                 - 1
                 - ((step - warmup_steps) // self.pp_group_size) % self.n_local_stages
             )
+            return (local_index * self.pp_group_size) + rank
+
+        fwd_stage_mb_index: Dict[int, int] = defaultdict(int)
+        bwd_stage_mb_index: Dict[int, int] = defaultdict(int)
+
+        # Store the list of operations used for that rank
+        rank_ops: List[Optional[Tuple[_ComputationType, int, int]]] = []
+        # Pre-padding, rank starts with no-ops based on the warmup.
+        for _ in range(rank):
+            rank_ops.append(None)
+
+        # These is used to calculate the number of slots to fill with no-ops, to account for the delay in warmup
+        # when we want to wait for the backward to trickle back up
+        # first backward is going to be pp_group_size - 1 hops away,
+        post_warmup_steps = 2 if rank == 0 else 0
```

Comment:
confused about this logic and the comment.

logic: why is rank 0 'special'?
comment: where does `pp_group_size - 1` actually factor in?

Reply:
Sorry the comment was outdated, I added the formula I used in the calculation in the comments.